torch/csrc/autograd/python_engine.cpp

```diff
@@ -33,38 +33,37 @@ namespace torch { namespace autograd { namespace python {
 PythonEngine::PythonEngine() = default;
 
 Engine& PythonEngine::get_python_engine() {
-  static PythonEngine engine;
+  // It's ok to "leak" PythonEngine singleton during application shutdown.
+  // Otherwise destructor needs to wait for all the worker/device threads to finish,
+  // which can be tricky as Python runtime was already destroyed at that point.
+  static PythonEngine *engine = new PythonEngine();
   // This is "probably" thread-safe because the flag is set in a fork handler
   // before any threads are created, and this function is only called with the
   // GIL held. However, using fork + threads is playing with fire so this is
   // more of a "best effort" thing. For example, if the fork occurs while the
   // backwards threads hold a lock, we'll probably deadlock in the engine
   // destructor.
   if (_reinitialize_engine) {
-    engine.release_workers();
-    engine.~PythonEngine();
-    new (&engine) torch::autograd::python::PythonEngine();
+    engine->release_workers();
+    delete engine;
+    engine = new PythonEngine();
     _reinitialize_engine = false;
   }
-  return engine;
+  return *engine;
 }
 
 void PythonEngine::thread_init(int device, const std::shared_ptr<ReadyQueue>& ready_queue, bool should_increment) {
-  // Increment thread usage count before acquiring the GIL
-  if (should_increment) {
-    increment_non_reentrant_thread_count();
-  }
+  //Ignore `should_increment` option, as PythonEngine device threads should never be destroyed
```

Comment:
I'm not sure about this.
In particular, after a fork, we want to properly clean and reset the threads.
The way it is done is pretend that no thread were actually attached and recreate them.

But if the whole logic to wait for the threads to die and count them is not used anymore, we should remote it right?

Reply:
I think it's still valuable to preserve the logic for PyTorch C++ (i.e. base `torch::autograd::Engine`) but kill it for `torch::autograd::python::PythonEngine`

Reply:
Ok.
I guess that's another bullet for the "rework autograd engine threading" task :D